Fix handling node types on JSDCE() in js-optimizer.js

gunyarakun opened the pull request:
This pull request fixes the problem reported in #4694, "Build fails on JSDCE() in js-optimizer.js with some special object keys."

kripken commented:
Nice, thank you!

The approach taken here is to check if something is quoted. I wonder though if we can do better, as something might be quoted but not be a reserved word, which is the underlying issue here, I think? Perhaps we should have a list of reserved words, and then we can still unquote the non-reserved? (Or does JSDCE not do unquoting anyhow, I don't remember? I guess you can just add a quoted non-reserved word to your test and see what happens.)

gunyarakun commented:
Thank you for your quick review and response!

The problem is not only the reserved words on JavaScript but also the reserved words on a UglifyJS AST. I added a test case with the word 'defun', which is not a reserved word in JavaScript because that seems internally used in UglifyJS or `tools/js-optimizer.js`. 

The logic I patched in JSDCE() is traversing all the AST and checking what functions are defined and used. The traversal is very simple just for detecting them and do not unquote anything. It cannot distinguish real function and the key of the Object 'function' currently because of the AST node expression is very simple in UglifyJS.

I tried to make as a small test and patch as possible. We could add other reserved words on JavaScript or UglifyJS AST into the test.

kripken commented:
Ok, sounds good. Please just add a quoted and non-quoted word that is non-reserved/non-special as a key to the new test. That way the test shows clearly how special words must be quoted while others do not need to be, and end up not quoted.

gunyarakun commented:
I modified and renamed the test case and realized that the hack just checking '.quoted' doesn't work well.
I wonder if you could keep this issue open and wait for the commit which fixes the problem

kripken commented:
Sure, no problem.

gunyarakun commented:
I made commits that JSDCE() checks the node is a variable name or an object key strictly.
With the commit 0950739f568be3823cd3267df0bf439f87b557c0, the tests in 7b3c9c08a50f14a1b1705e843467193d8057d563 work well.

kripken commented on tools/js-optimizer.js:
```diff
@@ -7769,6 +7769,8 @@ function eliminateDeadGlobals(ast) {
 // it goes away.
 function JSDCE(ast) {
   var scopes = [{}]; // begin with empty toplevel scope
+  var isVarNameOrObjectKey;
```
please add a comment explaining what this is and how it works

kripken replied:
Actually, the first could be moved into the functions, no need to declare it this high?  the second needs to stay here, and could benefit from a comment

gunyarakun replied:
Thank you for reviewing.
I'll move the variable declaration close to the place used and add a comment what is done how.

gunyarakun commented:
I added the commit corresponding to your comments.

kripken commented:
Thanks, looks great. Please just add yourself to AUTHORS before we merge.

gunyarakun commented:
Thank @kripken to review. I added myself into AUTHORS.
